clients/src/main/java/org/apache/kafka/common/protocol/SendBuilder.java

```diff
@@ -130,13 +139,25 @@ public void writeVarlong(long i) {
      */
     @Override
     public void writeRecords(BaseRecords records) {
-        flushCurrentBuffer();
-        sends.add(records.toSend(destinationId));
+        flushPendingBuffer();
+
+        if (records instanceof MemoryRecords) {
+            buffers.add(((MemoryRecords) records).buffer());
+        } else {
+            flushPendingSend();
+            sends.add(records.toSend(destinationId));
```

Comment:
How about setting initial size of ```sends``` to 1 as auto-generated code use only ```MemoryRecords```

Reply:
That's fair. `FetchResponse` is the only case where we would use a different `Records` type, so may as well optimize for the common case.